### PR TITLE
MULE-13217 Rework foreach to be non-blocking and decoupled from splitter

### PR DESCRIPTION
**Pull request by dfeist**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/api/routing/ForkJoinStrategy.java [hunk 1]
```diff
@@ -39,8 +39,20 @@
  */
 public interface ForkJoinStrategy {
 
+  /**
+   * Processes n routing pairs and aggregate the result as defined in the strategy implementation.
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/routing/ForkJoinStrategy.java]: n -> `{@code n}`
aggregate -> aggregates

File: core/src/main/java/org/mule/runtime/core/api/routing/ForkJoinStrategy.java [hunk 2]
```diff
@@ -39,8 +39,20 @@
...
+   * 
+   * @param original the original event
+   * @param routingPairs the routing paris to be
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/routing/ForkJoinStrategy.java]: to be ?

File: core/src/main/java/org/mule/runtime/core/internal/routing/Foreach.java [hunk 3]
```diff
@@ -60,98 +61,104 @@
...
+              responseBuilder.addVariable(counterVariableName, previousCounterVar);
             } else {
-              return e;
+              responseBuilder.removeVariable(counterVariableName);
             }
-          }).block();
-    } catch (Throwable throwable) {
-      throw rxExceptionToMuleException(throwable);
-    }
+            if (previousRootMessageVar != null) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/Foreach.java]: are this properties set again even in the case of errors? are there test cases for that scenario?

File: core/src/main/java/org/mule/runtime/core/internal/routing/Foreach.java [hunk 4]
```diff
@@ -60,98 +61,104 @@
...
-    Object payload = message.getPayload().getValue();
-    if (payload instanceof Document || payload.getClass().getName().startsWith("org.dom4j.")) {
-      return message;
-    } else {
-      return muleContext.getTransformationService().internalTransform(message, DataType.fromType(Document.class));
-    }
+  private Flux<InternalEvent> splitAndProcess(InternalEvent request) {
+    AtomicInteger count = new AtomicInteger();
+    return fromIterable(() -> splittingStrategy.split(request))
+        .transform(p -> batchSize > 1 ? from(p).buffer(batchSize).map(typedValueListToTypedValue()) : p)
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/Foreach.java]: explain with comments for normal humans to understand

File: core/src/main/java/org/mule/runtime/core/internal/routing/ScatterGatherRouter.java [hunk 5]
```diff
@@ -47,12 +48,15 @@
...
-        propagateWrappingFatal(new RoutePathNotFoundException(noEndpointsForRouter(), null));
-      }
-    };
+    return event -> validateMessageIsNotConsumable(event.getMessage());
+  }
+
+  @Override
+  public void initialise() throws InitialisationException {
+    super.initialise();
+    if (routes.size() < 2) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/ScatterGatherRouter.java]: previously only one route was valid. Is there any reason to don't allow that anymore?
I know it doesn't make much sense, just asking from a migration PoV

Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/internal/routing/ScatterGatherRouter.java]: This was a bug in my implementation, In 3.x minimum was two, also schema enforces 2.

File: core/src/main/java/org/mule/runtime/core/internal/routing/Foreach.java (hunk 3 above)
Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/internal/routing/Foreach.java]: Good point.  If an error with splitting occurs then this is not an issue but if error processing a part occurrs then need to take this into account, even if it was ignored in 3.x.

**Pull request merged**